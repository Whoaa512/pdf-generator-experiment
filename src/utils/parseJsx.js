import _ from 'lodash'
import { VM } from 'vm2'
import { transformAsync } from '@babel/core'
import JsxPdf from 'jsx-pdf'
import moment from 'moment-timezone'

import { log } from './log'

const DEFAULT_PARSE_TIMEOUT = 5000

export async function parseJsx(jsxString, timeout = DEFAULT_PARSE_TIMEOUT) {
    const code = `
        ;${jsxString};
        JsxPdf.renderPdf(doc)`

    if (!_.includes(jsxString, 'doc =')) {
        throw new Error('Missing required `doc` variable in template.')
    }

    const compiled = await transformAsync(code, {
        babelrc: false,
        plugins: [
            [
                '@babel/plugin-transform-react-jsx',
                {
                    pragma: 'JsxPdf.createElement',
                },
            ],
        ],
    })

    log.debug('~~~~~~~~')
    log.debug('transformed:', compiled.code)
    log.debug('~~~~~~~~')

    const vm = new VM({ timeout: timeout || DEFAULT_PARSE_TIMEOUT })

    vm.freeze(JsxPdf, 'JsxPdf')
    vm.freeze(moment, 'moment')

    return vm.run(`${compiled.code}`.replace('"use strict";', ''))
}
