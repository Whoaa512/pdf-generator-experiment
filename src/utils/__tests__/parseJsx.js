import fs from 'fs'
import path from 'path'

import pify from 'pify'

import { parseJsx } from '../parseJsx'

const readFile = pify(fs.readFile)

async function run(fileName) {
    const jsxString = await readFile(
        path.resolve(__dirname, '../../../test-fixtures/', fileName),
        { encoding: 'utf-8' },
    )

    return parseJsx(jsxString)
}

describe('parseJsx', () => {
    test('happy path', async () => {
        expect.assertions(1)

        const result = await run('test-jsx.txt')

        expect(result).toMatchSnapshot()
    })

    describe('invalid input', () => {
        test('attack-1', async () => {
            expect.assertions(1)

            const result = run('attack-1.txt')

            await expect(result).rejects.toThrowError(
                'Script execution timed out.',
            )
        })

        test('attack-2', async () => {
            expect.assertions(1)

            const result = run('attack-2.txt')

            await expect(result).rejects.toThrowError(
                `unknown: Unexpected token`,
            )
        })

        test('attack-3', async () => {
            expect.assertions(1)

            const result = run('attack-3.txt')

            await expect(result).rejects.toThrowError(
                'The <document> element can only contain <header>, <content>, and <footer> elements but found undefined',
            )
        })

        test('attack-4', async () => {
            expect.assertions(1)

            const result = run('attack-4.txt')

            await expect(result).rejects.toThrowError(
                "Cannot read property 'elementName' of null",
            )
        })

        test('attack-5', async () => {
            expect.assertions(1)

            const result = run('attack-5.txt')

            await expect(result).rejects.toThrowError(
                "Cannot read property 'call' of undefined",
            )
        })

        test('attack-6', async () => {
            expect.assertions(1)

            const result = run('attack-6.txt')

            await expect(result).rejects.toThrowError(
                'unknown: Unexpected token',
            )
        })

        test('attack-7', async () => {
            expect.assertions(1)

            const result = run('attack-7.txt')

            await expect(result).rejects.toThrowError(
                'unknown: Unexpected token',
            )
        })

        test('attack-8', async () => {
            expect.assertions(1)

            const result = run('attack-8.txt')

            await expect(result).rejects.toThrowError(
                'Script execution timed out.',
            )
        })

        test('attack-9', async () => {
            expect.assertions(1)

            const result = run('attack-9.txt')

            await expect(result).rejects.toThrowError('require is not defined')
        })
    })
})
