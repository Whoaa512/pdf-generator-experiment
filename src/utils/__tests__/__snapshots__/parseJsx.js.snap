// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseJsx happy path 1`] = `
Object {
  "content": Object {
    "stack": Array [
      "This will appear in my PDF!",
    ],
  },
  "defaultStyle": Object {
    "font": "OpenSans",
    "fontSize": 12,
  },
}
`;
